File: apps/backend/app/api/v1/project_clauses.py
"""Per-project Clause Library API.

Each project has its own clause library (project_clauses), built from that
project's uploaded contract. Routes are nested under a project:
    GET    /projects/{project_id}/clauses
    POST   /projects/{project_id}/clauses
    PATCH  /projects/{project_id}/clauses/{clause_id}
    DELETE /projects/{project_id}/clauses/{clause_id}
    POST   /projects/{project_id}/clauses/extract          (upload the contract)
    GET    /projects/{project_id}/clauses/extract-status   (poll AI extraction)

`extract` stores the uploaded contract in the project's data room and, when the
Anthropic key is configured, kicks off background AI extraction that fills the
project's clause library. The client polls `extract-status` for progress.
"""
import os
from datetime import datetime, timezone
from typing import List

from fastapi import APIRouter, BackgroundTasks, Depends, File, Form, HTTPException, UploadFile, status
from pydantic import BaseModel

from app.schemas.project_clause import (
    ProjectClauseCreate,
    ProjectClauseOut,
    ProjectClauseUpdate,
)
from app.services import clause_extraction, contract_book_service, pcc_comparison
from app.services.document_service import create_document, next_document_id
from app.services.project_clause_service import (
    add_clause,
    delete_clause,
    get_clause,
    get_clause_book_id,
    list_by_project,
    set_book_clauses,
    update_clause,
)
from app.api.v1.deps import get_current_user, require_permission

router = APIRouter()

# Adding/editing/removing clauses is a contract-management action — the same
# permission the global clause library used.
_MANAGE = require_permission("contracts.manage")

_EXT_TYPE = {
    "pdf": "PDF", "doc": "DOCX", "docx": "DOCX", "xls": "XLSX", "xlsx": "XLSX",
}


def _doc_type(name: str) -> str:
    ext = name.rsplit(".", 1)[-1].lower() if "." in name else ""
    return _EXT_TYPE.get(ext, "Other")


@router.get("/{project_id}/clauses", response_model=List[ProjectClauseOut])
async def get_project_clauses(project_id: str, _=Depends(get_current_user)):
    """This project's clause library (all authenticated users can read it)."""
    return list_by_project(project_id)


@router.post("/{project_id}/clauses", response_model=ProjectClauseOut, status_code=status.HTTP_201_CREATED)
async def create_project_clause(
    project_id: str,
    payload: ProjectClauseCreate,
    current_user=Depends(get_current_user),
    __=Depends(_MANAGE),
):
    if not payload.contract_standard.strip():
        raise HTTPException(status_code=400, detail="Contract standard is required.")
    if not payload.clause_number.strip():
        raise HTTPException(status_code=400, detail="Clause number is required.")
    if not payload.clause_title.strip():
        raise HTTPException(status_code=400, detail="Clause title is required.")
    return add_clause(project_id, payload.model_dump(), created_by=current_user["id"])


@router.patch("/{project_id}/clauses/{clause_id}", response_model=ProjectClauseOut)
async def patch_project_clause(
    project_id: str,
    clause_id: str,
    payload: ProjectClauseUpdate,
    _=Depends(get_current_user),
    __=Depends(_MANAGE),
):
    updated = update_clause(clause_id, payload.model_dump(exclude_none=True))
    if not updated:
        raise HTTPException(status_code=404, detail="Clause not found")
    return updated


@router.delete("/{project_id}/clauses/{clause_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_project_clause(
    project_id: str,
    clause_id: str,
    _=Depends(get_current_user),
    __=Depends(_MANAGE),
):
    if not get_clause(clause_id):
        raise HTTPException(status_code=404, detail="Clause not found")
    delete_clause(clause_id)


@router.post("/{project_id}/clauses/extract", status_code=status.HTTP_201_CREATED)
async def upload_contract_for_clauses(
    project_id: str,
    background: BackgroundTasks,
    file: UploadFile = File(...),
    current_user=Depends(get_current_user),
    __=Depends(_MANAGE),
):
    """Upload this project's contract and (when AI is configured) extract its clauses.

    The contract is stored in the project's data room, then Claude reads it in the
    background and fills the clause library. The client polls `extract-status`.
    If the Anthropic key isn't set, the contract is saved and the response says so.
    """
    content = await file.read()
    doc_id = next_document_id()
    record = {
        "id": doc_id,
        "projectId": project_id,
        "name": file.filename or "contract",
        "type": _doc_type(file.filename or ""),
        "sizeKB": max(1, round(len(content) / 1024)),
        "uploadedAt": datetime.now(timezone.utc).isoformat(),
        "uploadedBy": current_user.get("name") or current_user.get("email") or "",
        "status": "Uploaded",
        "claimRef": "Contract",
        "note": "Uploaded for clause extraction",
        "driveFileId": doc_id,
        "data": content,
        "mime": file.content_type or "application/octet-stream",
    }
    create_document(record)

    if not os.getenv("ANTHROPIC_API_KEY"):
        return {
            "stored": True,
            "documentId": doc_id,
            "status": "idle",
            "message": "Contract saved. AI clause extraction will run once the Anthropic key is enabled.",
        }

    clause_extraction.mark_running(project_id)
    background.add_task(clause_extraction.run_extraction, project_id, doc_id)
    return {
        "stored": True,
        "documentId": doc_id,
        "status": "running",
        "message": "Contract saved. Claude is reading it and extracting the clauses…",
    }


@router.get("/{project_id}/clauses/extract-status")
async def clause_extraction_status(project_id: str, _=Depends(get_current_user)):
    """Current state of background clause extraction for a project."""
    return clause_extraction.get_status(project_id)


# ── Base contract book (from the Knowledge Center) ───────────────────────────

class SelectBookPayload(BaseModel):
    bookId: str


@router.get("/{project_id}/clauses/book")
async def get_project_clause_book(project_id: str, _=Depends(get_current_user)):
    """The Knowledge-Center book chosen as this project's base clause set."""
    return {"bookId": get_clause_book_id(project_id)}


@router.post("/{project_id}/clauses/book")
async def select_project_clause_book(
    project_id: str,
    payload: SelectBookPayload,
    _=Depends(get_current_user),
    __=Depends(_MANAGE),
):
    """Pick a base FIDIC/standard-form book and copy its clauses into this
    project's library. Replaces the previously book-sourced clauses (manual and
    AI-extracted ones are kept) and clears any prior PCC modifications."""
    book = contract_book_service.get_book(payload.bookId)
    if not book:
        raise HTTPException(status_code=404, detail="Contract book not found.")
    if book.get("status") != "done":
        raise HTTPException(
            status_code=400,
            detail="This book hasn't finished extracting its clauses yet.",
        )
    book_clauses = contract_book_service.list_book_clauses(payload.bookId)
    if not book_clauses:
        raise HTTPException(status_code=400, detail="This book has no extracted clauses to copy.")
    count = set_book_clauses(project_id, book, book_clauses)
    # A new base book invalidates any earlier PCC comparison.
    pcc_comparison.mark_idle(project_id)
    return {"bookId": payload.bookId, "count": count}


# ── Particular Conditions of Contract (PCC) ─────────────────────────────────

@router.post("/{project_id}/clauses/pcc", status_code=status.HTTP_201_CREATED)
async def upload_pcc(
    project_id: str,
    background: BackgroundTasks,
    file: UploadFile = File(...),
    current_user=Depends(get_current_user),
    __=Depends(_MANAGE),
):
    """Upload the project's Particular Conditions of Contract. Claude compares it
    against the selected base book's clauses and flags the ones it amends.

    The PCC is stored in the project's data room, then compared in the background;
    the client polls `pcc-status`. Requires a base book to have been selected."""
    if not get_clause_book_id(project_id):
        raise HTTPException(
            status_code=400,
            detail="Select a base contract book before uploading the Particular Conditions.",
        )

    content = await file.read()
    doc_id = next_document_id()
    record = {
        "id": doc_id,
        "projectId": project_id,
        "name": file.filename or "particular-conditions",
        "type": _doc_type(file.filename or ""),
        "sizeKB": max(1, round(len(content) / 1024)),
        "uploadedAt": datetime.now(timezone.utc).isoformat(),
        "uploadedBy": current_user.get("name") or current_user.get("email") or "",
        "status": "Uploaded",
        "claimRef": "Particular Conditions",
        "note": "Uploaded for PCC comparison",
        "driveFileId": doc_id,
        "data": content,
        "mime": file.content_type or "application/octet-stream",
    }
    create_document(record)

    if not os.getenv("ANTHROPIC_API_KEY"):
        return {
            "stored": True,
            "documentId": doc_id,
            "status": "idle",
            "message": "Particular Conditions saved. AI comparison will run once the Anthropic key is enabled.",
        }

    pcc_comparison.mark_running(project_id)
    background.add_task(pcc_comparison.run_comparison, project_id, doc_id)
    return {
        "stored": True,
        "documentId": doc_id,
        "status": "running",
        "message": "Particular Conditions saved. Claude is comparing them against the base clauses…",
    }


@router.get("/{project_id}/clauses/pcc-status")
async def pcc_comparison_status(project_id: str, _=Depends(get_current_user)):
    """Current state of background PCC comparison for a project."""
    return pcc_comparison.get_status(project_id)

File: apps/web/src/api/projectClauses.ts
import { api } from "./client";
import type { ClauseRef } from "@/types";

/**
 * Raw clause row from the per-project `project_clauses` table. The card/table UI
 * works in terms of {@link ClauseRef} (book/clause/title/summary), so responses
 * are mapped to that shape here.
 */
interface ProjectClauseRow {
  id: string;
  projectId: string;
  contract_standard: string;
  clause_number: string;
  clause_title: string;
  clause_description: string;
  tags: string[];
  source?: string;
  modified?: boolean;
  modification_note?: string | null;
  created_by?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface ProjectClauseCreatePayload {
  contract_standard: string;
  clause_number: string;
  clause_title: string;
  clause_description?: string;
  tags?: string[];
}

export type ProjectClauseUpdatePayload = Partial<ProjectClauseCreatePayload>;

export interface ClauseExtractStatus {
  status: "idle" | "running" | "done" | "failed";
  error?: string;
  count?: number;
  /** PCC comparison: base clauses amended and brand-new clauses added. */
  modified?: number;
  added?: number;
}

export interface ContractExtractResult {
  stored: boolean;
  documentId: string;
  status: ClauseExtractStatus["status"];
  message: string;
}

function toClauseRef(row: ProjectClauseRow): ClauseRef {
  return {
    id: row.id,
    book: row.contract_standard,
    clause: row.clause_number,
    title: row.clause_title,
    summary: row.clause_description,
    tags: row.tags ?? [],
    source: row.source ?? "manual",
    modified: row.modified ?? false,
    modificationNote: row.modification_note ?? undefined,
  };
}

export async function listProjectClausesApi(projectId: string): Promise<ClauseRef[]> {
  const { data } = await api.get<ProjectClauseRow[]>(`/projects/${projectId}/clauses`);
  return data.map(toClauseRef);
}

export async function createProjectClauseApi(
  projectId: string,
  payload: ProjectClauseCreatePayload,
): Promise<ClauseRef> {
  const { data } = await api.post<ProjectClauseRow>(`/projects/${projectId}/clauses`, payload);
  return toClauseRef(data);
}

export async function updateProjectClauseApi(
  projectId: string,
  id: string,
  patch: ProjectClauseUpdatePayload,
): Promise<ClauseRef> {
  const { data } = await api.patch<ProjectClauseRow>(`/projects/${projectId}/clauses/${id}`, patch);
  return toClauseRef(data);
}

export async function deleteProjectClauseApi(projectId: string, id: string): Promise<void> {
  await api.delete(`/projects/${projectId}/clauses/${id}`);
}

/** Upload the project's contract; the backend extracts its clauses with AI in
 * the background. Returns quickly — poll {@link getClauseExtractStatusApi}. */
export async function extractProjectClausesApi(
  projectId: string,
  file: File,
): Promise<ContractExtractResult> {
  const form = new FormData();
  form.append("file", file);
  const { data } = await api.post<ContractExtractResult>(
    `/projects/${projectId}/clauses/extract`,
    form,
    { headers: { "Content-Type": "multipart/form-data" }, timeout: 60_000 },
  );
  return data;
}

/** Current state of background clause extraction for a project. */
export async function getClauseExtractStatusApi(projectId: string): Promise<ClauseExtractStatus> {
  const { data } = await api.get<ClauseExtractStatus>(`/projects/${projectId}/clauses/extract-status`);
  return data;
}

/** The Knowledge-Center book chosen as this project's base clause set (or null). */
export async function getClauseBookApi(projectId: string): Promise<string | null> {
  const { data } = await api.get<{ bookId: string | null }>(`/projects/${projectId}/clauses/book`);
  return data.bookId;
}

/** Pick a base contract book; its clauses are copied into the project's library. */
export async function selectClauseBookApi(
  projectId: string,
  bookId: string,
): Promise<{ bookId: string; count: number }> {
  const { data } = await api.post<{ bookId: string; count: number }>(
    `/projects/${projectId}/clauses/book`,
    { bookId },
  );
  return data;
}

/** Upload the project's Particular Conditions; the backend compares them against
 * the base book's clauses in the background. Poll {@link getPccStatusApi}. */
export async function uploadPccApi(
  projectId: string,
  file: File,
): Promise<ContractExtractResult> {
  const form = new FormData();
  form.append("file", file);
  const { data } = await api.post<ContractExtractResult>(
    `/projects/${projectId}/clauses/pcc`,
    form,
    // A PCC can be a large (scanned) file — the comparison itself runs in the
    // background, but the upload of the bytes needs a generous timeout.
    { headers: { "Content-Type": "multipart/form-data" }, timeout: 120_000 },
  );
  return data;
}

/** Current state of the background PCC comparison for a project. */
export async function getPccStatusApi(projectId: string): Promise<ClauseExtractStatus> {
  const { data } = await api.get<ClauseExtractStatus>(`/projects/${projectId}/clauses/pcc-status`);
  return data;
}
